File: DevelopArchitecture/DevelopArchitecture/threeclass/JWhenTapped/JWhenTappedView.m
//
//  JWhenTappedView.m
//  DevelopArchitecture
//
//  Created by Jeffrey on 2020/1/16.
//  Copyright © 2020 Jeffrey. All rights reserved.
//

#import "JWhenTappedView.h"
#import <objc/runtime.h>

@interface UIView (JWhenTappedViewBlocks_Private)

- (void)funj_runBlockForKey:(void *)blockKey;
- (void)funj_setBlock:(JWhenTappedViewBlock)block forKey:(void *)blockKey;

- (UITapGestureRecognizer*)funj_addTapGestureRecognizerWithTaps:(NSUInteger) taps touches:(NSUInteger) touches selector:(SEL) selector;
- (void) funj_addRequirementToSingleTapsRecognizer:(UIGestureRecognizer*) recognizer;
- (void) funj_addRequiredToDoubleTapsRecognizer:(UIGestureRecognizer*) recognizer;

@end

@implementation UIView (JWhenTappedView)

static char kWhenTappedBlockKey;
static char kWhenDoubleTappedBlockKey;
static char kWhenLongPressedBlockKey;
static char kWhenTouchedDownBlockKey;
static char kWhenTouchedUpBlockKey;

#pragma mark -
#pragma mark Set blocks

- (void)funj_runBlockForKey:(void *)blockKey {
    JWhenTappedViewBlock block = objc_getAssociatedObject(self, blockKey);
    if (block) block(self);
}

- (void)funj_setBlock:(JWhenTappedViewBlock)block forKey:(void *)blockKey {
    self.userInteractionEnabled = YES;
    objc_setAssociatedObject(self, blockKey, block, OBJC_ASSOCIATION_COPY_NONATOMIC);
}

#pragma mark -
#pragma mark When Tapped

- (void)funj_whenTapped:(JWhenTappedViewBlock)block {
    UITapGestureRecognizer* gesture = [self funj_addTapGestureRecognizerWithTaps:1 touches:1 selector:@selector(funj_viewWasTapped)];
    [self funj_addRequiredToDoubleTapsRecognizer:gesture];
    
    [self funj_setBlock:block forKey:&kWhenTappedBlockKey];
}

- (void)funj_whenDoubleTapped:(JWhenTappedViewBlock)block {
    UITapGestureRecognizer* gesture = [self funj_addTapGestureRecognizerWithTaps:2 touches:1 selector:@selector(funj_viewWasDoubleTapped)];
    [self funj_addRequirementToSingleTapsRecognizer:gesture];
    
    [self funj_setBlock:block forKey:&kWhenDoubleTappedBlockKey];
}

- (void)funj_whenLongPressed:(JWhenTappedViewBlock)block{
     UILongPressGestureRecognizer *longPress =[[UILongPressGestureRecognizer alloc]initWithTarget:self action:@selector(funj_viewWasLongPressed)];
    [self addGestureRecognizer:longPress];
    
    [self funj_setBlock:block forKey:&kWhenLongPressedBlockKey];
}

- (void)funj_whenTouchedDown:(JWhenTappedViewBlock)block {
    [self funj_setBlock:block forKey:&kWhenTouchedDownBlockKey];
}

- (void)funj_whenTouchedUp:(JWhenTappedViewBlock)block {
    [self funj_setBlock:block forKey:&kWhenTouchedUpBlockKey];
}

#pragma mark -
#pragma mark Callbacks

- (void)funj_viewWasTapped {
    [self funj_runBlockForKey:&kWhenTappedBlockKey];
}

- (void)funj_viewWasDoubleTapped {
    [self funj_runBlockForKey:&kWhenDoubleTappedBlockKey];
}

- (void)funj_viewWasLongPressed {
    [self funj_runBlockForKey:&kWhenLongPressedBlockKey];
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesBegan:touches withEvent:event];
    [self funj_runBlockForKey:&kWhenTouchedDownBlockKey];
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
    [super touchesEnded:touches withEvent:event];
    [self funj_runBlockForKey:&kWhenTouchedUpBlockKey];
}

#pragma mark -
#pragma mark Helpers

- (UITapGestureRecognizer*)funj_addTapGestureRecognizerWithTaps:(NSUInteger)taps touches:(NSUInteger)touches selector:(SEL)selector {
    UITapGestureRecognizer* tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:selector];
    tapGesture.delegate = self;
    tapGesture.numberOfTapsRequired = taps;
    tapGesture.numberOfTouchesRequired = touches;
    [self addGestureRecognizer:tapGesture];
    
    return tapGesture;
}

- (void) funj_addRequirementToSingleTapsRecognizer:(UIGestureRecognizer*) recognizer {
    for (UIGestureRecognizer* gesture in [self gestureRecognizers]) {
        if ([gesture isKindOfClass:[UITapGestureRecognizer class]]) {
            UITapGestureRecognizer* tapGesture = (UITapGestureRecognizer*) gesture;
            if (tapGesture.numberOfTouchesRequired == 1 && tapGesture.numberOfTapsRequired == 1) {
                [tapGesture requireGestureRecognizerToFail:recognizer];
            }
        }
    }
}

- (void) funj_addRequiredToDoubleTapsRecognizer:(UIGestureRecognizer*) recognizer {
    for (UIGestureRecognizer* gesture in [self gestureRecognizers]) {
        if ([gesture isKindOfClass:[UITapGestureRecognizer class]]) {
            UITapGestureRecognizer* tapGesture = (UITapGestureRecognizer*) gesture;
            if (tapGesture.numberOfTouchesRequired == 2 && tapGesture.numberOfTapsRequired == 1) {
                [recognizer requireGestureRecognizerToFail:tapGesture];
            }
        }
    }
}

@end

File: DeveloperArchitecture/DeveloperArchitecture/threeclass/whenTapped/JWhenTappedView.swift
//
//  JWhenTappedView.swift
//  DeveloperArchitecture
//
//  Created by Jeffrey on 2020/10/30.
//  Copyright © 2020 Jeffery. All rights reserved.
//

import Foundation
import UIKit

private var kWhenTappedBlockKey = "WhenTappedBlockKey"
private var kWhenDoubleTappedBlockKey = "WhenDoubleTappedBlockKey"
private var kWhenLongPressedBlockKey = "WhenLongPressedBlockKey"
private var kWhenTouchedDownBlockKey = "WhenTouchedDownBlockKey"
private var kWhenTouchedUpBlockKey = "WhenTouchedUpBlockKey"

typealias JWhenTappedViewBlock = ((_ view : UIView)->())

extension UIView {
    func funj_whenTapped(_ block :@escaping JWhenTappedViewBlock){
        let gesture = UITapGestureRecognizer(target: self, action: #selector(funj_viewWasTapped))
        self.addGestureRecognizer(gesture)
        self.m_whenTapped = block
    }
    func funj_whenDoubleTapped(_ block :@escaping JWhenTappedViewBlock){
        let gesture = UITapGestureRecognizer(target: self, action: #selector(funj_viewWasDoubleTapped))
        gesture.numberOfTapsRequired = 2
        self.addGestureRecognizer(gesture)
        self.m_whenDoubleTapped = block
    }
    func funj_whenLongPressed(_ block :@escaping JWhenTappedViewBlock){
        let gesture = UILongPressGestureRecognizer(target: self , action: #selector(funj_viewWasLongPressed))
        self.addGestureRecognizer(gesture)
        self.m_whenLongPressed = block
    }
    func funj_whenTouchedDown(_ block :@escaping JWhenTappedViewBlock){
        self.m_whenTouchedDown = block
    }
    func funj_whenTouchedUp(_ block :@escaping JWhenTappedViewBlock){
        self.m_whenTouchedUp = block
    }
}
extension UIView {
    var m_whenTapped : JWhenTappedViewBlock? {
        get { return objc_getAssociatedObject(self, &kWhenTappedBlockKey) as? JWhenTappedViewBlock }
        set {objc_setAssociatedObject(self, &kWhenTappedBlockKey, newValue, .OBJC_ASSOCIATION_COPY_NONATOMIC)}}
    var m_whenDoubleTapped : JWhenTappedViewBlock? {
        get { return objc_getAssociatedObject(self, &kWhenDoubleTappedBlockKey) as? JWhenTappedViewBlock }
        set {objc_setAssociatedObject(self, &kWhenDoubleTappedBlockKey, newValue, .OBJC_ASSOCIATION_COPY_NONATOMIC)}}
    var m_whenLongPressed : JWhenTappedViewBlock? {
        get { return objc_getAssociatedObject(self, &kWhenLongPressedBlockKey) as? JWhenTappedViewBlock }
        set {objc_setAssociatedObject(self, &kWhenLongPressedBlockKey, newValue, .OBJC_ASSOCIATION_COPY_NONATOMIC)}}
    var m_whenTouchedDown : JWhenTappedViewBlock? {
        get { return objc_getAssociatedObject(self, &kWhenTouchedDownBlockKey) as? JWhenTappedViewBlock }
        set {objc_setAssociatedObject(self, &kWhenTouchedDownBlockKey, newValue, .OBJC_ASSOCIATION_COPY_NONATOMIC)}}
    var m_whenTouchedUp : JWhenTappedViewBlock? {
        get { return objc_getAssociatedObject(self, &kWhenTouchedUpBlockKey) as? JWhenTappedViewBlock }
        set {objc_setAssociatedObject(self, &kWhenTouchedUpBlockKey, newValue, .OBJC_ASSOCIATION_COPY_NONATOMIC)}}
    
    @objc func funj_viewWasTapped(){
        self.m_whenTapped?(self)
    }
    @objc func funj_viewWasDoubleTapped(){
        self.m_whenDoubleTapped?(self)
    }
    @objc func funj_viewWasLongPressed(){
        self.m_whenLongPressed?(self)
    }
    open override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        self.m_whenTouchedUp?(self)
    }
    open override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesEnded(touches, with: event)
        self.m_whenTouchedDown?(self)
    }
}
